### DISCUSSION TITLE
Better way to filter by date / time ?

### DISCUSSION THREAD

**Discussion by cduytran** [pocketbase/pocketbase#1791] [Q&A] [answered]
I am in (UTC-8:00) Pacific Time (US & Canada). I am trying to query data from today's date. This is what I have done so far.

```
let currDate = new Date( )
currDate.setHours( currDate.getHours( ) - 8 )

let startDate = currDate.toISOString().slice(0,10)

let beginTime = new Date(startDate+"T00:00:00")
let stopTime = new Date(startDate+"T23:59:00")
  
beginTime = beginTime.toISOString().replace(/T/," ")
stopTime = stopTime.toISOString().replace(/T/," ")

const activeRecords = await locals.pb.collection('pack').getFullList( 200,{
    filter:`updated >= "${beginTime}" && updated <= "${stopTime}"`,
  })
```

So my logics are : 
- Offset -8 hours to convert my local time to ISO time
- Extract the yyyy-mm-dd from the converted date
- Create two new dates representing starting time and stopping time
- Replace the T from ISOString with a blank ( " ") because pocketbase date doesn't have the T in the middle
- Finally execute query

Is there a better way to implement this? 

Note: I just started programming for 3 weeks.. so take it easy on me. Thanks.

**Comment by ganigeorgiev** [comment]
You don't have to manually offset your local date since [`toISOString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString) per spec should always return the date in UTC. Also note that `new Date(startDate+"T00:00:00")` will result to a new date with 00:00:00 time in your local timezone because it is missing a suffix (for UTC it is `Z`).

**Edit: the below is not entirely correct since it will target only the UTC day range, which is not what the author of the post want. See the [comment below for more details](https://github.com/pocketbase/pocketbase/discussions/1791#discussioncomment-4881369)**.

~~Here is an alternative version:~~

```js
const today = new Date( ).toISOString().slice(0,10)
const beginTime = today + " 00:00:00.000"
const stopTime = today + " 23:59:59.999"

const activeRecords = await locals.pb.collection('pack').getFullList( 200,{
    filter:`updated >= "${beginTime}" && updated <= "${stopTime}"`,
})

// OR for a simple day range you can also just use `~` operator:
const activeRecords = await locals.pb.collection('pack').getFullList( 200,{
    // updated ~ "2023-02-06"
    filter:`updated ~ "${today}"`,
})
```

**Comment by cduytran** [comment, marked as answer]
Thanks for the quick reply.

I don't think the second solution works for my scenario. 

Let's say I have a record with updated value '2023-02-06 02:49:32.788Z' which is 2/5/2023, 6:49:32 PM my local time

`const today = new Date( ).toISOString().slice(0,10)` will return 2023-02-06

If I run query with filter:`updated ~ "${today}"` , the result would return that record.

The query should only return results with my local time which is 2/6/2023.

Man, this is kinda confusing... :(

**Reply by cduytran** [reply]
My ultimate goal is to get all records that have been updated for today (my local time).

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what do you mean. The dates in PocketBase are stored only in UTC and so the filters must be also in UTC.

`new Date( ).toISOString().slice(0,10)` will return the current **local** date as UTC. If you want to use a timezone different then the local one of your browser, then you can manually offset the date as in your initial example.

**Reply by cduytran** [reply]
I am confused too.. :-) .. basically I have a collection called pack. At the end of the day, I'd like to get all the pack(s) that have been updated for that day. So any pack that was updated yesterday shouldn't appear on my query.  I can't follow the UTC time zone because it's 8 hours ahead of me.

I am stuck :-( ... i am so confused with the UTC  time zone.

**Reply by ganigeorgiev** [reply]
> I can't follow the UTC time zone because it's 8 hours ahead of me.

`toISOString()` will return your current local browser time into a UTC one. You can check your current browser timezone via `console.log(Intl.DateTimeFormat().resolvedOptions().timeZone)`.
If you don't want to rely on the client browser tz, you can use a fixed offset as in your initial example.

**Reply by cduytran** [reply]
Ok, I think I can explain myself now. By the way, thanks for your patient.

Lets say I have one record updated at 2023-02-06 10:00:00 UTC which is 2023-02-06 02:00:00 my local time

If I run this right now
`const today = new Date( ).toISOString().slice(0,10)`  // 2023-02-06

const activeRecords = await locals.pb.collection('pack').getFullList( 200,{
    // updated ~ "2023-02-06"
    filter:`updated ~ "${today}"`,
})

This code should return one record.

And let's say later , I created another record at 2023-02-06 18:00:00 my local time which is 2023-02-07 06:00:00 UTC.
This second record will have 2023-02-07 06:00:00 UTC as the updated value.

Then I run this the above code again; 

`const today = new Date( ).toISOString().slice(0,10)`  <--` today` is now 2023-02-07 

and  this code
> const activeRecords = await locals.pb.collection('pack').getFullList( 200,{
    // updated ~ "2023-02-06"
    filter:`updated ~ "${today}"`,
})
 will not include the first record; which is what we don't want.

Man, it's late... :-( ... i cant sleep

**Reply by ganigeorgiev** [reply]
Ah, I understand now what you meant earlier. In this case, yes, my example will not work since it will target only the UTC day.

If you want to provide a day range within your local timezone you'll have to convert it as in your example, I don't think there is other way. A slightly shorter alternative could be:

```js
let beginTime = new Date();
beginTime.setHours(0, 0, 0, 0);
beginTime = beginTime.toISOString().replace("T", " ")

let stopTime = new Date();
stopTime.setHours(23, 59, 59, 999);
stopTime = stopTime.toISOString().replace("T", " ")

const activeRecords = await locals.pb.collection("pack").getFullList(200, {
    filter:`updated >= "${beginTime}" && updated <= "${stopTime}"`,
})
```

**Reply by cduytran** [reply]
This works beautifully. Thanks so much. Now I can sleep.

I appreciate all the help. You are the BESTTTTT !!!!!

Man, programming can be so addicting and sometimes frustrating as well.

Is javascript ( specially svelte ) worth learning? I am just starting out. And I am loving it. Gonna replace my membrane keyboard for a mechanical keyboard... my fingers are hurting from all the typing...
